src/mlpack/methods/ann/layer/sigmoid_cross_entropy_error.hpp

```diff
@@ -0,0 +1,120 @@
+/**
+ * @file sigmoid_cross_entropy_error.hpp
+ * @author Kris Singh
+ *
+ * Definition of the cross-entropy with logit performance function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_SIGMOID_CROSS_ENTROPY_ERROR_HPP
+#define MLPACK_METHODS_ANN_LAYER_SIGMOID_CROSS_ENTROPY_ERROR_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The SigmoidCrossEntropyError performance function measures the network's
+ * performance according to the cross-entropy function between the input and
+ * target distributions. This function calculates the cross entropy
+ * given the real values instead of providing the sigmoid activations.
+ * The functions is much more numerically stable.
+ *
+ * For brevity, let x = logits, z = labels. The logistic loss is:
+ * \f$ z * -\log(sigmoid(x)) + (1 - z) * -\log(1 - sigmoid(x))\f$
+ * \f$= z * -\log(1 / (1 + e^{-x})) + (1 - z) * -\log(e^{-x} / (1 + e^{-x}))\f$
+ * \f$= z * \log(1 + e^{-x}) + (1 - z) * (-\log(e^{-x}) + \log(1 + e^{-x}))\f$
+ * \f$= z * \log(1 + e^{-x}) + (1 - z) * (x + \log(1 + e^{-x})\f$
+ * \f$= (1 - z) * x + \log(1 + e^{-x})\f$
+ * \f$= x - x * z + \log(1 + e^{-x})\f$
+ *
+ * For x < 0, to avoid overflow in e^{-x}, we reformulate the above:
+ * \f$x - x * z + \log(1 + e^{-x})\f$
+ * \f$= \log(e^{x}) - x * z + \log(1 + e{-x})\f$
+ * \f$= - x * z + \log(1 + e^{x})\f$
+ *
+ * Hence, to ensure stability and avoid overflow, the implementation uses this
+ * equivalent formulation:
+ * \f$max(x, 0) - x * z + \log(1 +  e^{-|x|})\f$
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
```

Comment:
The details actually follow from [here](https://github.com/mlpack/mlpack/pull/1104#discussion_r137058610). I was unable to find a paper to cite for the above.

Reply:
I know, looks like I missed that in summer.

Reply:
I think it would be fine to drop derivation and just use: ``f(x,t) = x - x * t + log(1 + exp(-x))`` where ``x`` is the input and ``t`` are the target values.

At the end it's just Sigmoid + CrossEntropy nothing special, if you like you can cite "On Loss Functions for Deep Neural Networks in Classification" by K. Janocha et al.

Reply:
@rcurtin had previously advocated for an optimized implementation. I'd like to wait and see if he concurs with the same.

Reply:
In the context of this comment I think it's just fine to take @zoq's suggestion.  No need for a full derivation in the comments, in my view.